File: backend/src/services/user.service.ts
import bcrypt from 'bcryptjs';
import { userProvider } from '../providers/userProvider';
import { AppError } from '../utils/AppError.js';
import {
  UserCreateBodyType,
  UserResType,
  UserRes,
  UserUpdateBodyType,
} from '../schemas/user';
import { BaseListResType, BaseSearchRequestType } from '../schemas/search';
import { roleProvider } from '../providers/roleProvider';

export const userService = {
  async create(data: UserCreateBodyType): Promise<UserResType> {
    var exstingEmail = await userProvider.findByEmail(data.email);
    if (exstingEmail) throw new AppError(409, 'Email already registered');

    const hashed = await bcrypt.hash(data.password, 12);
    var result = await userProvider.create({ ...data, password: hashed });
    const role = await roleProvider.findRoleByName('Customer'.toUpperCase());
    if (role) await roleProvider.assignRole(result.id, role.id);
    return UserRes.parse(result);
  },

  async update(data: UserUpdateBodyType, id: number) {
    var user = await userProvider.findById(id);
    if (!user) throw new AppError(404, 'User not found');

    var result = await userProvider.update(data, id);
    return UserRes.parse(result);
  },

  async findById(id: number) {
    var result = await userProvider.findById(id);
    return UserRes.parse(result);
  },

  async findByPhone(phone: string) {
    var result = await userProvider.findByPhone(phone);
    return UserRes.parse(result);
  },

  async delete(id: number) {
    var user = await userProvider.findById(id);
    if (!user) throw new AppError(404, 'User not found');
    var result = await userProvider.delete(id);
    return UserRes.parse(result);
  },

  async findAll(request: BaseSearchRequestType): Promise<BaseListResType> {
    const limit = request.limit;
    const users = await userProvider.findAll(request);
    const hasNextPage = users.length > limit;
    const items = hasNextPage ? users.slice(0, limit) : users;
    const nextCursor =
      hasNextPage && items.length > 0 ? items[items.length - 1].id : null;
    return {
      data: items.map((user: any) => UserRes.parse(user)),
      limit,
      nextCursor,
      hasNextPage,
    };
  },

  async assignRole(userId: number, roleId: number) {
    var user = await userProvider.findById(userId);
    if (!user) throw new AppError(404, 'User not found');

    var role = await roleProvider.findById(roleId);
    if (!role) throw new AppError(404, 'Role not found');

    await roleProvider.assignRole(user.id, role.id);
    return true;
  },

  async removeAssignRole(userId: number, roleId: number) {
    var user = await userProvider.findById(userId);
    if (!user) throw new AppError(404, 'User not found');

    var role = await roleProvider.findById(roleId);
    if (!role) throw new AppError(404, 'Role not found');

    await roleProvider.removeAssignRole(user.id, role.id);
    return true;
  },

  async updateAvatar(userId: number, imgUrl: string): Promise<UserResType> {
    var user = await userProvider.findById(userId);
    if (!user) throw new AppError(404, 'User not found');

    var result = await userProvider.updateImg(userId, imgUrl);
    return UserRes.parse(result);
  },
};

File: backend/app/services/user_service.py
from sqlalchemy.ext.asyncio import AsyncSession
from app.providers.user_provider import user_provider
from app.providers.role_provider import role_provider
from app.utils.app_error import AppError
from app.utils.auth_utils import hash_password
from typing import Optional

async def find_all(db: AsyncSession, limit: int = 10, cursor: Optional[int] = None, order: str = "desc", search: Optional[str] = None) -> dict:
    rows = await user_provider.find_all(db, limit=limit, cursor=cursor, order=order, search=search)
    has_next = len(rows) > limit
    items = rows[:limit] if has_next else rows
    next_cursor = items[-1].id if has_next and items else None
    return {"data": items, "limit": limit, "nextCursor": next_cursor, "hasNextPage": has_next}

async def find_by_id(db: AsyncSession, user_id: int):
    user = await user_provider.find_by_id(db, user_id)
    if not user:
        raise AppError(404, "User not found")
    return user

async def find_by_phone(db: AsyncSession, phone: str):
    user = await user_provider.find_by_phone(db, phone)
    if not user:
        raise AppError(404, "User not found")
    return user

async def create(db: AsyncSession, dto: dict):
    existing = None
    if dto.get("email"):
        existing = await user_provider.find_by_email(db, dto["email"])
    if existing:
        raise AppError(409, "Email already registered")
    data = {k: v for k, v in dto.items() if v is not None}
    if data.get("password"):
        data["password"] = hash_password(data["password"])
    return await user_provider.create(db, **data)

async def update(db: AsyncSession, dto: dict, user_id: int):
    data = {k: v for k, v in dto.items() if v is not None}
    user = await user_provider.update(db, user_id, **data)
    if not user:
        raise AppError(404, "User not found")
    return user

async def delete(db: AsyncSession, user_id: int):
    user = await user_provider.soft_delete(db, user_id)
    if not user:
        raise AppError(404, "User not found")
    return user

async def update_avatar(db: AsyncSession, user_id: int, img: str):
    user = await user_provider.update_avatar(db, user_id, img)
    if not user:
        raise AppError(404, "User not found")
    return user

async def assign_role(db: AsyncSession, user_id: int, role_id: int) -> bool:
    return await role_provider.assign_role(db, user_id, role_id)

async def remove_assign_role(db: AsyncSession, user_id: int, role_id: int) -> bool:
    return await role_provider.remove_role(db, user_id, role_id)
